Docker + CI Updates

cocosolos opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Adds Docker support. Images are built and published on commits to base. See [the README](https://github.com/cocosolos/server/blob/docker/docker/README.md) for more details.

~~Adds a development container. See [the README](https://github.com/cocosolos/server/blob/docker/.devcontainer/README.md) for more details.~~

Updates CI. Breaks the build and test jobs into their own workflow files. Create unique build caches on commits to base, ensuring CI runs always use an up to date build cache. Conditionally run clang-tidy only on PRs that include cpp changes. Fail runs when any builds/sanity checks fail.

Reorganize sanity checks scripts. Running `./tools/ci/sanity_checks.sh` will run all of the sanity checks, comparing your current changes to `origin/base`. You can change the branch this is compared to by setting the `GIT_REF` env var (e.g. `GIT_REF=upstream/base ./tools/ci/sanity_checks.sh`). Output goes into `sanity_checks_summary.md` and is formatted for better readability.

## Steps to test these changes

[Example cache building run](https://github.com/cocosolos/server/actions/runs/18280108033)
[Example passing CI run](https://github.com/cocosolos/server/actions/runs/17715629138)

Images are available for testing, replace `ghcr.io/LandSandBoat` with `ghcr.io/cocosolos` in the commands in the Docker README.

cocosolos committed:
Embed debug information in Windows executables

Using external debugging info causes issues with relative paths.

cocosolos committed:
Add Docker support

Ubuntu and Alpine based images, with options for a GCC or Clang build.
Caching optimized for local building.
See docker/README.md for more info.

cocosolos committed:
Update CI/CD

Build cache on commits to base using proper cache keys.
Use fresh cache during CI builds (big improvement to build times).
Skip clang-tidy if no source changes during CI builds.
Improve CI results using GitHub Actions summaries.
Move CI test job into separate workflow.
Combine Windows and Linux test workflows.

Note:
ccache cache isn't being extracted from Docker image, causing full
rebuilds every time. Disable Docker build in CI workflows until we can
fix this.

cocosolos committed:
Add local development compose file

cocosolos committed:
Update fmt

Fixes issue with newer clang version:
fmt/format.h:4429:27: error: identifier '_a' preceded by whitespace in a literal operator declaration is deprecated [-Werror,-Wdeprecated-literal-operator]

cocosolos committed:
Format Python files with Black formatter

zach2good commented on cmake/StandardProjectSettings.cmake:
```diff
@@ -66,11 +66,15 @@ if(MSVC)
             /GR # Enable RTTI
         )
+        string(REPLACE "/Zi" "/Z7" CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG}")
+        string(REPLACE "/Zi" "/Z7" CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG}")
```
You've already defined `CMAKE_MSVC_DEBUG_INFORMATION_FORMAT` and set the `CMP0141` policy, so manually setting the flags should be irrelevant now? Though I believe policies have to be defined before the call to `project()` at the top level. I'm not sure when `CMAKE_MSVC_DEBUG_INFORMATION_FORMAT ` wants to be called.

We want to avoid this pattern in cmake: `string(REPLACE `, it's quite sensitive (as I'm discovering with my cmake rewrite). 


https://cmake.org/cmake/help/latest/policy/CMP0141.html
https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_DEBUG_INFORMATION_FORMAT.html

cocosolos replied:
Will look into this. This was a quick fix after some googling a while ago so probably some unnecessary stuff here.

cocosolos replied:
Cleaned this up and moved to the top of root CMakeLists.txt.

zach2good commented on dev.docker-compose.yml:
```diff
@@ -0,0 +1,213 @@
+x-dbcreds: &dbcreds
```
Can this (and all docker-related files) live in `docker/`?

zach2good replied:
and what does the `x-` mean here?

cocosolos replied:
> and what does the `x-` mean here?

https://docs.docker.com/reference/compose-file/extension/


> Can this (and all docker-related files) live in `docker/`?

I think dockerignore needs to be in the root context. The compose file can probably be moved, will look into it.

cocosolos replied:
Okay so moving `dev.docker-compose.yml` into `docker/`:

If you want to run `docker compose up` without specifying the file or any other cli args (e.g. `docker compose -f docker/docker-compose.yml up`) you need a file named `docker-compose.yml` in the root context. The idea behind `dev.docker-compose.yml` is that you can use it as-is straight from the vscode editor window or copy it into the untracked `docker-compose.yml` to make edits and/or run it in a terminal without adding extra args.

Moving `dev.docker-compose.yml` would require changing the bind points to mount the parent directory. If you then want to customize this or otherwise use the services from the terminal as mentioned above, you'd no longer be able to simply copy/paste the dev file, you'd need to make edits to all of the bind points. I think this change would unnecessarily complicate the process.

zach2good replied:
> If you want to run docker compose up without specifying the file or any other cli args (e.g. docker compose -f docker/docker-compose.yml up) you need a file named docker-compose.yml in the root context.

For me, this is one of the most basic things I want from the whole dockerization process, so here it shall stay 👍

zach2good commented on .devcontainer/README.md:
```diff
@@ -0,0 +1,7 @@
+# Dev Container
```
Dev container is a cool idea, but please pop this commit off and do it after the main docker stuff lands - to bring the scope of this PR down

cocosolos replied:
Dropped.

zach2good commented on .vscode/extensions.json:
```diff
@@ -2,7 +2,7 @@
         "sumneko.lua",
-        "wwm.better-align",
+        "Chouzz.vscode-better-align",
         "tomblind.local-lua-debugger-vscode",
```
I believe the lua debugger is also deprecated and will warn

cocosolos replied:
Went ahead and removed this and tws to match #8323.

zach2good commented on .github/workflows/publish_meshes.yml:
```diff
@@ -0,0 +1,53 @@
+name: publish_meshes
```
Really, should this be happening from the respective losmeshes and navmeshes repos?

cocosolos replied:
I'm not really sure. Losmeshes isn't controlled by LSB so I figured we just update it whenever the submodules get updated. It would also complicate it a little since you'd need to pull both images. Could clone the latest from the parent repos but then triggering it becomes an issue.

zach2good commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,182 @@
+
+      - name: Install Python dependencies
+        run: |
+          python3 -m venv .venv
+          source ${{ runner.os == 'Windows' && '.venv/Scripts/activate' || '.venv/bin/activate' }}
```
It really was as simple as that to do venv, cool 👍

zach2good commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,41 @@
+                    "terminal.integrated.defaultProfile.linux": "bash",
+                    "terminal.integrated.profiles.linux": { "bash": { "path": "/bin/bash" } }
+                },
+                "extensions": [
+                    "sumneko.lua",
+                    "Chouzz.vscode-better-align",
+                    "ms-python.python",
+                    "ms-python.black-formatter",
+                    "eamodio.gitlens",
+                    "donjayamanne.githistory",
+                    "josetr.cmake-language-support-vscode",
+                    "ms-vscode.cpptools-themes",
+                    "ms-vscode.cpptools-extension-pack",
+                    "ms-vscode.cpptools",
+                    "jeff-hykin.better-cpp-syntax",
+                    "EditorConfig.EditorConfig",
+                    "github.vscode-github-actions",
+                    "bierner.github-markdown-preview",
+                    "xaver.clang-format"
+                ]
+            }
```
I wonder if this can be made to read or mount what we have in `.vscode` so we don't have to keep two of this list

cocosolos replied:
So this file actually installs all of these extensions while `.vscode` is just recommendations. I think if this is excluded then you have to manually install the recommendations each time the container is created which diminishes some of the "out of the box" experience of the devcontainer and loses out on some caching/persistence.

zach2good commented on .devcontainer/README.md:
```diff
@@ -0,0 +1,7 @@
+# Dev Container
+
+The dev container uses the devtools image as a base for a full-featured development environment. Using the dev container will provide you with all of our needed dependencies in an isolated environment as well as a preconfigured VS Code with our recommended settings and extensions for development, formatting, and testing.
```
The concept is fantastic, but is this skirting the "mounting host filesystem stuff in local(WSL) is painfully slow and you should recreate in the local docker filesystem"?

Is there syncing of changes from (what I assume is) the local filesystem back to the host system? Otherwise this will feel quite counterintuitive to work on (where are the changes I just made?)

Or is it meant to be/feel entirely self-contained?

cocosolos replied:
This bind mounts your current workspace into the container, so you're working on your local files but using the docker environment for dependencies. So when you edit files you're directly editing your local files. All output from working inside the devcontainer (like built executables, logs,  and the build directory) will also appear on your host filesystem.

If you're working on Windows you should have the code checked out into the WSL filesystem. Note this doesn't mean open WSL and launch vscode, but actually have LSB checked out in the WSL VM for example in your WSL home directory. Unfortunately Windows isn't an ideal environment to run this in. It should work and it _shouldn't_ cause too much overhead (other than the WSL VM itself). But Docker is a Linux focused technology and there are some issues that exist with Docker on Windows.

zach2good replied:
I think there's no way around it: we'll just have to provide the big fat warning in the future docs section that talks about docker saying: "Jesus Christ, you must do it this way, or the experience will be unusable"

zach2good replied:
Or... detect that it's happening? Put that idea on a "nice to haves in the future" list

zach2good commented on .editorconfig:
```diff
@@ -79,3 +79,7 @@ indent_brace_style = allman # This is unfortunately ignored in Visual Studio
 [Makefile]
 indent_style = tab
+
+[*.yml]
+indent_style = space
+indent_size = 2
```
This will be useful to start thinking about for possible future data storage: what is the industry standard here? This is good because I've accidentally re-formatted our CI files at least twice

cocosolos replied:
I think 2 spaces is the standard.

zach2good commented on .github/workflows/build_docker.yml:
```diff
@@ -0,0 +1,178 @@
+          labels: ${{ steps.docker-meta.outputs.labels }}
+          annotations: ${{ steps.docker-meta.outputs.annotations }}
+          platforms: linux/amd64
+          build-args: |
+            COMPILER=${{ inputs.compiler }}
+            CMAKE_BUILD_TYPE=${{ inputs.build_type }}
+            TRACY_ENABLE=${{ inputs.tracy && 'ON' || 'OFF' }}
+            PCH_ENABLE=${{ inputs.build_type == 'Debug' && 'OFF' || 'ON' }}
+            REPO_URL=${{ github.server_url }}/${{ github.repository }}.git
+            COMMIT_SHA=${{ github.sha }}
+          outputs: ${{ inputs.upload_artifact && format('type=docker,dest={0}/LSB_{1}_image_{2}.tar', runner.temp, inputs.os, github.sha) }}
```
👍

zach2good commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,324 @@
+  Build:
+    needs: Sanity_Checks
+    name: ${{ format('Build ({0})', matrix.runner) }}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - runner: windows-2025
+            compiler: msvc19
+          - runner: macos-15
+            compiler: appleClang16
+          - runner: ubuntu-24.04
+            compiler: gcc14
```
👍

zach2good commented on .github/workflows/codeql_analysis.yml:
```diff
@@ -47,6 +47,7 @@ jobs:
     - name: Initialize CodeQL
       uses: github/codeql-action/init@v2
       with:
+        trap-caching: false
```
What's this do?

cocosolos replied:
This stops codeql from saving a ~1.2gb cache to the github actions cache that can cause storage limit issues with build caching storage. I couldn't figure out what it's actually used for but it doesn't seem to have any effect on the codeql action completion time and I don't think it affects the results. If it's needed it can be added back but we'd need to be careful about it bumping off one of the build cache entries, though that may require some specific circumstances to happen. Mostly just to be safe.

zach2good commented on .github/workflows/publish_latest.yml:
```diff
@@ -0,0 +1,88 @@
+name: publish_latest
```
What are the circumstances that might cause this to fail? (we blow through our storage, etc.)

What happens to us if this job fails? etc.

cocosolos replied:
If you go over the storage limit GH will just start deleting older cache entries at some point. It's not a big deal and it'll manage itself. If this job fails it just won't create the cache and/or won't publish a new `latest` image. CI runs will try to use an older cache if it exists or will just run uncached.

zach2good commented on dev.docker-compose.yml:
```diff
@@ -0,0 +1,213 @@
+  MARIADB_PASSWORD: 'password'
+  MARIADB_RANDOM_ROOT_PASSWORD: true
+
+x-common: &common
```
As a docker noob, what does this mean?

cocosolos replied:
Just some things that are used for a few different services, so split out so it's easier to maintain. See the link above about extensions.

cocosolos replied:
```
  image: landsandboat/server:testing
  pull_policy: never
  ```
  `image` is just the local image name, can be whatever you want (copy/paste into an untracked docker-compose if you wanna edit). `pull_policy` just says not to try to remotely pull any image that matches that name.

zach2good commented on dev.docker-compose.yml:
```diff
@@ -0,0 +1,213 @@
+
+services:
+  database:
+    image: mariadb:lts
```
This is great, if we can ditch mine or clays altered actions for CI

cocosolos replied:
Unfortunately this only works for Linux.

zach2good commented on dev.docker-compose.yml:
```diff
@@ -0,0 +1,213 @@
+      search:
+        condition: service_started
+
+volumes:
+  losmeshes:
+    external: true
+  navmeshes:
+    external: true
```
But this can be empty or these _don't have to_ be provided? We should probably just enforce the use of all meshes on all platforms - it shouldn't be optional anymore

cocosolos replied:
Basically the same if you don't checkout submodules. Def don't want to include them in the image since they're huge, but maybe map can warn/exit if meshes aren't found.

zach2good commented on docker/alpine.Dockerfile:
```diff
@@ -0,0 +1,231 @@
+########
+# Base #
+########
+FROM alpine:3.22 AS base
```
I assume all the deps and stuff in here are version locked because of this (good)?

cocosolos replied:
Yeah this is their "stable" release version. `edge` is their rolling release. Looks like they cut a release twice a year, and the next one should be in May (about 1 month after next ubuntu lts).

zach2good commented on docker/alpine.Dockerfile:
```diff
@@ -0,0 +1,231 @@
+    if grep -qF 'u_int32_t' $EFSW_FILE; then
+        echo "Patching $EFSW_FILE: Replacing u_int32_t with uint32_t"
+        sed -i 's/u_int32_t/uint32_t/g' $EFSW_FILE
+    fi
```
I can just fix/update these deps for you?

cocosolos replied:
Whatever you prefer. This was another one of those early hack fixes.

cocosolos replied:
Rebased to pull in #8372, dropped this manual patching. Thanks for taking care of that!

cocosolos replied:
...and apparently touched `ext/CMakeLists.txt` again. Will see if I can fix that.

cocosolos replied:
I think this is actually the right line endings. `git add --renormalize .` changes that file, so it seems it was committed with incorrect line endings somehow (despite `.gitattributes`, which should supersede `core.autocrlf` if set).

https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration#_formatting_and_whitespace
https://git-scm.com/docs/git-add#Documentation/git-add.txt---renormalize

cocosolos committed:
Retry certain CI jobs on failure

Windows startup checks and potentially APT can be problematic.

cocosolos commented:
Rebased to resolve a conflict and added a new commit that reruns some problematic workflow steps. Startup checks occasionally stops responding (mostly Windows from what I've seen), so give it 10 mins to do its thing with 3 attempts. APT also apparently has had issues in the past, so 5 mins should be enough for that (though I haven't seen this in my testing).

Workflow run where Windows startup checks fail twice before completing the third attempt:
https://github.com/cocosolos/server/actions/runs/18551464372
